### PR TITLE
Use proper shell for modal dialogs

### PR DESCRIPTION
**Pull request by iloveeclipse**
Fixes https://github.com/eclipse-platform/eclipse.platform/issues/1373

### CODE REVIEW

File: debug/org.eclipse.debug.ui/ui/org/eclipse/debug/internal/ui/sourcelookup/Prompter.java [hunk 1]
```diff
@@ -78,11 +90,59 @@ public Object handleStatus(final IStatus status, final Object source) throws Cor
...
 		if (exception[0] != null ) {
 			throw exception[0];
 		}
 		return result[0];
 	}
+
+	/**
+	 * Helper to activate possible opened and hidden prompt dialog
+	 */
+	private final class ForcePrompShellActivation implements Runnable {
```
Comment (trancexpress) [inline debug/org.eclipse.debug.ui/ui/org/eclipse/debug/internal/ui/sourcelookup/Prompter.java:105]: This can be `static`.

**Comment by iloveeclipse** [comment]
The reason I've added the change in the Prompter is that there is not only one handler that can open dialogs, it is defined by extension point and anyone can do whatever possible. So I would prefer to have a more generic solution.

The "borderless" dialog is the result of `SWT.ON_TOP` flag? Looks quite ... unusual :-)
Is there a more "convenient looking" way to achieve the same effect?

**Comment by tomaswolf** [comment]
Wait a sec. Why exactly is this happening?

The progress dialog comes from `IProgressService.busyCursorWhile()`, which is defined not to open a progress dialog if there is a modal dialog. The "Errors in Workspace" dialog _is_ a modal dialog. And showing a dialog _must_ happen on the UI thread, right? So there can be exactly two outcomes:
* Error dialog comes first, and then no progress dialog is shown (yet).
* Or progress dialog comes first, and then error dialog is shown on top.

_However_: I do notice that the error dialog opening in `CompileErrorProjectPromptStatusHandler` uses the active workbench window's shell as parent for its message dialog. But modal dialogs should always use `PlatformUI.getWorkbench().getModalDialogShellProvider().getShell()`. I suspect #1373 is actually case (2): progress dialog comes first, but then error dialog comes later _but has the wrong parent shell_ and thus ends up somehow behind the progress dialog. (Besides, having two modal dialogs open simultaneously from the same parent shell quite often results in UI deadlocks in my experience.)

**Comment by iloveeclipse** [comment]
> So there can be exactly two outcomes:

I believe there are three, and third one is: progress dialog is opened **after** error dialog.

**Comment by iloveeclipse** [comment]
> > So there can be exactly two outcomes:
> 
> I believe there are three, and third one is: progress dialog is opened **after** error dialog.

Or four: focus / activation is given to the wrong shell - because of timing issues in SWT event handling, I believe I saw some SWT code about "fix shell focus" in GTK port.

**Comment by tomaswolf** [comment]
> > So there can be exactly two outcomes:
> 
> I believe there are three, and third one is: progress dialog is opened **after** error dialog.

If so, the `IProgressService.busyCursorWhile()` is broken. Its javadoc clearly states that it does not open the dialog if there is already a modal dialog.

https://github.com/eclipse-platform/eclipse.platform.ui/blob/deb75d4e977d767d3c5249e8a953fd2ef986baab/bundles/org.eclipse.ui.workbench/Eclipse%20UI/org/eclipse/ui/progress/IProgressService.java#L132-L133

I would really suggest to at least just try changing `CompileErrorProjectPromptStatusHandler` to use  `PlatformUI.getWorkbench().getModalDialogShellProvider().getShell()` instead of `DebugUIPlugin.getShell()` at https://github.com/eclipse-platform/eclipse.platform/blob/6cb771edfb0680235d2a371668f8e9d194458fba/debug/org.eclipse.debug.ui/ui/org/eclipse/debug/internal/ui/launchConfigurations/CompileErrorPromptStatusHandler.java#L41

**Comment by iloveeclipse** [comment]
I will play tomorrow with the code, sure.

**Comment by iloveeclipse** [comment]
> I would really suggest to at least just try changing `CompileErrorProjectPromptStatusHandler` to use `PlatformUI.getWorkbench().getModalDialogShellProvider().getShell()` instead of `DebugUIPlugin.getShell()`

This works perfectly, many thanks Thomas for the hint! Now progress dialog can't even be activated. I will change related handlers to that.

**Comment by merks** [comment]
This is such a good example of excellent teamwork!

**Comment by iloveeclipse** [comment]
I've touched all `IStatusHandler` implementations in platform, the rest (JDT/PDE) will follow with similar changes.

**Comment by trancexpress** [comment]
The change seems way more straightforward now.

File: debug/org.eclipse.debug.ui/ui/org/eclipse/debug/internal/ui/DebugUIPlugin.java [hunk 2]
```diff
@@ -372,6 +372,22 @@ public static Shell getShell() {
...
+	 * dialog, or, if there is no one, currently active workbench window shell or
+	 * <code>null</code> if none.
+	 *
+	 * @return the modal shell or currently active workbench window shell or
+	 *         <code>null</code>
+	 */
+	public static Shell getShellForModalDialog() {
+		Shell shell = PlatformUI.getWorkbench().getModalDialogShellProvider().getShell();
+		if (shell == null) {
+			shell = getShell();
```
Comment (tomaswolf) [inline debug/org.eclipse.debug.ui/ui/org/eclipse/debug/internal/ui/DebugUIPlugin.java:386]: Why is this needed? Seems to me that the (default) IShellProvider `ProgressManagerUtil::getDefaultParent` returns a shell if there is one. If it returns null, there is no shell, and the `DebugUIPlugin` won't find any either.

**Comment by iloveeclipse** [comment]
@tomaswolf , @trancexpress : any objections? If not, I would merge.

File: debug/org.eclipse.debug.ui/ui/org/eclipse/debug/internal/ui/DebugUIPlugin.java (hunk 2 above)
Comment (tomaswolf) [reply debug/org.eclipse.debug.ui/ui/org/eclipse/debug/internal/ui/DebugUIPlugin.java:386]: Yes, it can return null. My point was: If it returns null, then `getShell()` won't find anything either.

**Comment by tomaswolf** [comment]
No objections, but I still don't see the point of calling `DebugUIPlugin.getShell()` as a "fallback". It won't hurt, though.

Thanks!
